### PR TITLE
Bugfix/5487 auto lr ordering

### LINKED ISSUE #5487: auto learning rate finder fails AFTER optimization due to misconfiguration
## 🐛 Bug
Optimizing learning rate using auto_lr_find=True fails **AFTER** a learning rate was found with error

`pytorch_lightning.utilities.exceptions.MisconfigurationException: When auto_lr_find is set to True, expects that `model` or `model.hparams` either has field `lr` or `learning_rate` that can overridden`

**AFTER** having waited for several minutes.

This can be caught **BEFORE** the long run to save time.

## Please reproduce using the BoringModel

I can just copy paste everything to here... the only lacking line is `self.lr = 999` under `model::__init__()`, if it is not present the error happens, otherwise it doesn't happen.

Reproducing this a single time takes me 15 minutes (due to the lr finding process), but I still wanted to report this because it seems important and I don't have time right now for a full reproduction on the template.

`
trainer = Trainer(auto_lr_find=True)
trainer.tune(model)
`

then for a while I see the progress bar

`Finding best initial lr:  27%|██▋       | 27/100 [02:43<07:37,  6.27s/it]`

and then the error after a lr has been found.

### Expected behavior

`pytorch_lightning.utilities.exceptions.MisconfigurationException: When auto_lr_find is set to True, expects that `model` or `model.hparams` either has field `lr` or `learning_rate` that can overridden`

should hit BEFORE the long calculation.

### Environment

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): Windows 
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): -
 - Python version: 3.7
 - CUDA/cuDNN version: 101
 - GPU models and configuration: single GPU NVIDIA quadro P4000
 - Any other relevant information:-

### PR DESCRIPTION
**Pull request by noamzilo**
## What does this PR do?

before: auto learning rate finder fails AFTER optimization due to misconfiguration
after: auto learning rate finder fails BEFORE optimization due to misconfiguration, and user can fail faster.

Fixes #5487

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by noamzilo** [comment]
To the maintainers:

I have added a test that requires `pip install pytest-timeout`


Currently, I am getting the error

_______________ ERROR collecting tests/trainer/test_lr_finder.py _______________
'timeout' not found in `markers` configuration option
_______________ ERROR collecting tests/trainer/test_lr_finder.py _______________
'timeout' not found in `markers` configuration option

Will it be possible to install this library so the test will pass?

If not, please recommend what I should use instead so I will be able to change the code.

**Comment by carmocca** [comment]
> test that requires pip install pytest-timeout

why is it required?

If you are just changing the order of the check, we probably don't need a test for it. Usually you want to test your changes but there's no need to test the call order. And if there was, you could do it by mocking the called functions and checking the order.

File: pytorch_lightning/tuner/lr_finder.py [hunk 1]
```diff
@@ -40,8 +40,28 @@
     from tqdm import tqdm
 
 
+def __choose_lr_assigner(trainer, model: LightningModule) -> Callable[[Any], None]:
```
Comment (carmocca) [inline pytorch_lightning/tuner/lr_finder.py:43]: Instead of returning the `Callable`, can you return the attribute string and then only call `lightning_setattr` in `_run_lr_finder_internally`?

**Comment by noamzilo** [comment]
> 
> 
> > test that requires pip install pytest-timeout
> 
> why is it required?
> 
> If you are just changing the order of the check, we probably don't need a test for it. Usually you want to test your changes but there's no need to test the call order. And if there was, you could do it by mocking the called functions and checking the order.

actually the call order that wasn't testes is what is PR is about. I could just delete the test... should I do that? I think this tests that the user experience is such that no waiting for errors can occur.

About the code comment - will commit the str return soon.

**Comment by Borda** [comment]
@noamzilo @carmocca how is it going here?

File: pytorch_lightning/tuner/lr_finder.py (hunk 1 above)
Comment (noamzilo) [reply pytorch_lightning/tuner/lr_finder.py]: fixed

**Comment by noamzilo** [comment]
On CI complete testing / pytest (ubuntu-18.04, 3.6, minimal) I am getting 

'timeout' not found in `markers` configuration option

Please tell me if I should remove the test. I believe that changing it to just test of ordering doesn't make sense. I think pytest-timeout should be added.

If not, I will delete the test and be done with it.

**Comment by carmocca** [comment]
Hi @noamzilo. I have simplified the test for you. You can check if tests what you wanted.

**Comment by carmocca** [comment]
Also, can you update the CHANGELOG?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/tuner/lr_finder.py [hunk 2]
```diff
@@ -42,34 +42,36 @@
...
+        return trainer.auto_lr_find
+
+    if lightning_hasattr(model, 'lr'):
+        return 'lr'
+    if lightning_hasattr(model, 'learning_rate'):
+        return 'learning_rate'
+    raise MisconfigurationException(
+        'When auto_lr_find is set to True, expects that `model` or'
+        ' `model.hparams` either has field `lr` or `learning_rate`'
+        ' that can overridden')
```
Comment (Borda) [inline pytorch_lightning/tuner/lr_finder.py:60]:
```suggestion
    attr_options = ('lr', 'learning_rate')
    for attr in attr_options:
        if lightning_hasattr(model, attr):
            return attr

    raise MisconfigurationException(
        'When auto_lr_find is set to True, expects that `model` or'
        f' `model.hparams` either has one of these fields {attr_options}'
        ' that can overridden')
```

File: tests/trainer/test_lr_finder.py [hunk 3]
```diff
@@ -262,3 +264,12 @@ def test_suggestion_with_non_finite_values(tmpdir):
...
     assert before_lr == after_lr, \
         'Learning rate was altered because of non-finite loss values'
+
+
+def test_lr_finder_fails_fast_on_bad_config(tmpdir):
+    """ Test that tune fails if the model does not have a lr BEFORE running lr find """
+    # note: this did not raise an exception before #5648 because lr_find is skipped
+    # during fast_dev_run and the lr attribute check was done after lr_find
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, auto_lr_find=True)
+    with pytest.raises(MisconfigurationException, match='either has field `lr`'):
```
Comment (Borda) [inline tests/trainer/test_lr_finder.py:274]:
```suggestion
    with pytest.raises(MisconfigurationException, match='either has one of these fields'):
```

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/trainer/test_lr_finder.py [hunk 4]
```diff
@@ -262,3 +264,12 @@ def test_suggestion_with_non_finite_values(tmpdir):
 
     assert before_lr == after_lr, \
         'Learning rate was altered because of non-finite loss values'
+
+
+def test_lr_finder_fails_fast_on_bad_config(tmpdir):
+    """ Test that tune fails if the model does not have a lr BEFORE running lr find """
+    # note: this did not raise an exception before #5648 because lr_find is skipped
+    # during fast_dev_run and the lr attribute check was done after lr_find
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, auto_lr_find=True)
```
Comment (rohitgr7) [inline tests/trainer/test_lr_finder.py:273]: with `fast_dev_run=True`, lr_find is disabled, how is it generating this error here?

Comment (rohitgr7) [reply tests/trainer/test_lr_finder.py:273]: should tuner be enabled with fast_dev_run? cc @SkafteNicki

File: pytorch_lightning/tuner/lr_finder.py [hunk 5]
```diff
@@ -42,34 +42,37 @@
 log = logging.getLogger(__name__)
 
 
+def __determine_lr_attr_name(trainer, model: LightningModule) -> str:
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/lr_finder.py:45]:
```suggestion
def _determine_lr_attr_name(trainer, model: LightningModule) -> str:
```
any reason for this?

File: pytorch_lightning/tuner/lr_finder.py [hunk 6]
```diff
@@ -42,34 +42,37 @@
...
+
+    raise MisconfigurationException(
+        'When auto_lr_find is set to True, expects that `model` or'
+        f' `model.hparams` either has one of these fields {attr_options}'
+        ' that can overridden')
+
+
 def _run_lr_finder_internally(trainer, model: LightningModule):
     """ Call lr finder internally during Trainer.fit() """
+    lr_attr_name = __determine_lr_attr_name(trainer, model)
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/lr_finder.py:66]:
```suggestion
    lr_attr_name = _determine_lr_attr_name(trainer, model)
```
if yes pls ignore

File: tests/trainer/test_lr_finder.py [hunk 7]
```diff
@@ -262,3 +264,12 @@ def test_suggestion_with_non_finite_values(tmpdir):
 
     assert before_lr == after_lr, \
         'Learning rate was altered because of non-finite loss values'
+
+
+def test_lr_finder_fails_fast_on_bad_config(tmpdir):
+    """ Test that tune fails if the model does not have a lr BEFORE running lr find """
+    # note: this did not raise an exception before #5648 because lr_find is skipped
+    # during fast_dev_run and the lr attribute check was done after lr_find
```
Comment (rohitgr7) [inline tests/trainer/test_lr_finder.py:272]: #5648 is an issue, how does it relate here?

File: tests/trainer/flags/test_fast_dev_run.py [hunk 8]
```diff
@@ -15,6 +15,7 @@ def test_skip_on_fast_dev_run_tuner(tmpdir, tuner_alg):
     """ Test that tuner algorithms are skipped if fast dev run is enabled """
 
     model = BoringModel()
+    model.lr = 0.1  # avoid no-lr-found exception
```
Comment (rohitgr7) [inline tests/trainer/flags/test_fast_dev_run.py:18]: pls override the `BoringModel` wherever `lr` is required

File: pytorch_lightning/tuner/lr_finder.py [hunk 9]
```diff
@@ -42,34 +42,37 @@
 log = logging.getLogger(__name__)
 
 
+def __determine_lr_attr_name(trainer, model: LightningModule) -> str:
+    if isinstance(trainer.auto_lr_find, str):
+        if not lightning_hasattr(model, trainer.auto_lr_find):
+            raise MisconfigurationException(
+                f'`auto_lr_find` was set to {trainer.auto_lr_find}, however'
+                ' could not find this as a field in `model` or `model.hparams`.')
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/lr_finder.py:50]:
```suggestion
            raise MisconfigurationException(
                f'`auto_lr_find` was set to {trainer.auto_lr_find}, however'
                ' could not find this as a field in `model` or `model.hparams`.'
            )
```

File: pytorch_lightning/tuner/lr_finder.py [hunk 10]
```diff
@@ -42,34 +42,37 @@
...
+
+    attr_options = ('lr', 'learning_rate')
+    for attr in attr_options:
+        if lightning_hasattr(model, attr):
+            return attr
+
+    raise MisconfigurationException(
+        'When auto_lr_find is set to True, expects that `model` or'
+        f' `model.hparams` either has one of these fields {attr_options}'
+        ' that can overridden')
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/lr_finder.py:61]:
```suggestion
    raise MisconfigurationException(
        'When `auto_lr_find=True`, either `model` or `model.hparams` should'
        f' have one of these fields: {attr_options}', that can be overridden.'
    )
```

**Review by rohitgr7 — APPROVED** [review]

File: tests/trainer/test_lr_finder.py (hunk 4 above)
Comment (SkafteNicki) [reply tests/trainer/test_lr_finder.py:273]: the intention is that tuner should be disabled during fast_dev_run

Comment (carmocca) [reply tests/trainer/test_lr_finder.py:273]: Agree with @SkafteNicki 

The point of this test is to ensure that the lr attribute check is done before running lr_find

File: tests/trainer/test_lr_finder.py (hunk 7 above)
Comment (carmocca) [reply tests/trainer/test_lr_finder.py:272]: typo, meant to reference this PR

```suggestion
    # note: this did not raise an exception before #5638 because lr_find is skipped
    # during fast_dev_run and the lr attribute check was done after lr_find
```

File: tests/trainer/flags/test_fast_dev_run.py (hunk 8 above)
Comment (carmocca) [reply tests/trainer/flags/test_fast_dev_run.py:18]: This is fine since the attribute is set on the variable, not the class

File: pytorch_lightning/tuner/lr_finder.py [hunk 11]
```diff
@@ -42,34 +42,38 @@
...
+        return trainer.auto_lr_find
+
+    attr_options = ('lr', 'learning_rate')
+    for attr in attr_options:
+        if lightning_hasattr(model, attr):
+            return attr
+
+    raise MisconfigurationException(
+        'When `auto_lr_find=True`, either `model` or `model.hparams` should'
+        f' have one of these fields: {attr_options}', that can be overridden.'
```
Comment (carmocca) [inline pytorch_lightning/tuner/lr_finder.py:61]:
```suggestion
        f' have one of these fields: {attr_options} overridden.'
```

File: tests/trainer/test_lr_finder.py [hunk 12]
```diff
@@ -262,3 +264,12 @@ def test_suggestion_with_non_finite_values(tmpdir):
...
     assert before_lr == after_lr, \
         'Learning rate was altered because of non-finite loss values'
+
+
+def test_lr_finder_fails_fast_on_bad_config(tmpdir):
+    """ Test that tune fails if the model does not have a lr BEFORE running lr find """
+    # note: this did not raise an exception before #5638 because lr_find is skipped
+    # during fast_dev_run and the lr attribute check was done after lr_find
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, auto_lr_find=True)
+    with pytest.raises(MisconfigurationException, match='either has one of these fields'):
```
Comment (carmocca) [inline tests/trainer/test_lr_finder.py:274]:
```suggestion
    with pytest.raises(MisconfigurationException, match='should have one of these fields'):
```

File: tests/trainer/test_lr_finder.py (hunk 4 above)
Comment (rohitgr7) [reply tests/trainer/test_lr_finder.py]: but for batch_size_scaling, its opposite. It checks only if it's going to run it but here it first checks and then decide whether to run.

Comment (rohitgr7) [reply tests/trainer/test_lr_finder.py]: ok I see, there are two different method calls for lr_find but only one for batch_size_caling:

Comment (rohitgr7) [reply tests/trainer/test_lr_finder.py]: can we have a check for `fast_dev_run` inside `_run_lr_finder_internally` too? just to have the same behavior in both?

Comment (SkafteNicki) [reply tests/trainer/test_lr_finder.py]: i think that is a great idea @rohitgr7 !

Comment (rohitgr7) [reply tests/trainer/test_lr_finder.py]: cool, let's do in a follow up PR. I have a better suggestion for this.

**Review by rohitgr7 — APPROVED** [review]
